[WIP] New Feature: Add database functionality

WassCodeur opened the pull request:
### Add database functionality

#### Description
This PR adds database functionality to manage data and its menus. The main changes are as follows:

- Added a `restaurant_menus` collection to store information about restaurants and their menus.
- Added verification to avoid duplicates before inserting documents.
- Use of the `pymongo` module to interact with MongoDB.

#### Tests carried out
- Verification of document insertion without duplication.
- Tested basic CRUD operations on the `restaurant_menus` collection.

#### How to test
1. Install dependencies with `pip install -r requirements.txt`.
2. Configure  [MongoDB](https://www.prisma.io/dataguide/mongodb/setting-up-a-local-mongodb-database) and launch the server.
3. Create the `.env` file, copy the contents of `.env.example` and update the database variables with the values from your mongodb database: 
4. Run the `data/db.py` script to test functionality.

#### Additional comments
We welcome any suggestions or improvements!

WassCodeur committed:
NF: Add database functionality

koviubi56 reviewed (changes requested):
1. Setup instructions for the db should be included in the README.
2. Better docstrings for classes and functions. (see [here](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html))
3. Follow PEP8 by using snake_case for function names.
4. Don't return strings as a result; raise an exception for errors.
5. Simplify if-statements:
  ```py
  # instead of this:
  if self.connected:
     # ...code...
  else:
     raise ...

  # do this:
  if not self.connected:
      raise ...
  # ...code...
  ```
6. `dict.get(x)` might return None. Either set a default or use `dict[x]`
7. Try to use general type hints for parameters:
  ```py
  # x doesn't need to be a list, it just needs to be iterable
  def f(x: Iterable[str]) -> None:
      for something in x:
          g(something)
  ```
8. Consider raising an exception instead of returning None.
9. I'm not sure if we're gonna have tests.

WassCodeur commented:
Hi @koviubi56 !

Thank you for your feedback

koviubi56 reviewed (approved):
It LGTM, but I'm wondering: is it worth it to use MongoDB? You have to install it, set it up; while there is [`sqlite3`](https://docs.python.org/3/library/sqlite3.html) in stdlib that doesn't need to be set up. So is there a specific reason you chose MongoDB?

WassCodeur commented:
Hello @koviubi56,

thanks for your comment. In fact, I used mongoDB because of its ability to support several types of data (dynamic schema and rich queries) for one entity, in addition to its scalability and high read/write performance. By the way, you can also use the online version of mongoDB, so there's no need to install the server locally.

Vinyzu commented:
This is marked as a WIP (Work in Progress).
The current goal of the database is a MongoDB which we can push/pull the savestates to.
We might extend the databases use case if we feel like it fits us well.

WassCodeur committed:
[FT]: Add functionality to save and load game states

koviubi56 commented on requirements.txt:
```diff
@@ -3,6 +3,4 @@
 disnake~=2.9.2
 python-dotenv~=1.0.1
 dacite~=1.8.1
-torch~=2.3.1
-sentence-transformers~=3.0.1
-numpy<2
```
Why did these get removed?

koviubi56 commented on src/sincere_singularities/data/db.py:
```diff
@@ -0,0 +1,225 @@
+        """
+        if not self.connected:
+            raise ConnectError("Not connected to the database")
+        if self.show_one(collection, data):
+            self.db[collection].delete_one(data)
```
Maybe raise an exception if it didn't exist in the first place? (also in `delete_many`)

koviubi56 commented on src/sincere_singularities/data/db.py:
```diff
@@ -0,0 +1,225 @@
+        if self.show_one(collection, data):
+            self.db[collection].delete_one(data)
+
+    def show_all(self, collection: str) -> Iterable[dict[str, Any]]:
```
You can specify that the return value is a list.

koviubi56 commented on src/sincere_singularities/data/db.py:
```diff
@@ -0,0 +1,225 @@
+        element = self.db[collection].find_one(data)
+        if element:
+            return dict(element)
+        return {}
```
Raise instead.

koviubi56 commented on src/sincere_singularities/data/db.py:
```diff
@@ -0,0 +1,225 @@
+
+        element = self.show_one(collection, data)
+
+        if element:
```
Raise exception if it doesn't exist in the first place. (also in update_many)

koviubi56 commented on src/sincere_singularities/data/restaurants.json:
```diff
@@ -1,4 +1,52 @@
 [
```
Add points to the restaurants. The first one (index 0) should need 0 points, and the following ones should have more and more (see Pizzaria and Sushi Restaurant for example).

koviubi56 replied:
Actually, we already worked on this, so this addition isn't necessary.

koviubi56 commented on src/sincere_singularities/data/savestates.py:
```diff
@@ -0,0 +1,98 @@
+        if not self.client.is_connected():
+            raise ConnectError("Not connected to the database")
+
+    def add_user_state(self, data: dict[str, Any]) -> None:
```
Maybe you could make this a TypedDict

koviubi56 commented on src/sincere_singularities/data/states.json:
```diff
@@ -0,0 +1,63 @@
+[
```
Thanks, but I don't think we need example states.

koviubi56 reviewed (changes requested):
Thank you. If you could fix my recommendations, that'd be great. Also, could you implement what I temporarily implemented in `points.py` (between the two comments)?

Vinyzu committed:
Order Conditions (#14)

* Order Conditions

- Implemented Various Order Conditions
- Improved Naming
- Small Fixes

Co-Authored-By: koviubi56 <koviubi56@duck.com>

* Fix Debug-Statements, Problems

---------

Co-authored-by: koviubi56 <koviubi56@duck.com>

Vinyzu committed:
Various Game Improvements (#15)

WassCodeur committed:
chore: Update .gitignore and requirements.txt, and remove unused files

Update .gitignore to include json files for test cases. Also, update requirements.txt to fix the missing newline at the end of the file. Additionally, remove the unused states.json file.

WassCodeur committed:
chore: Update order.py and restaurant.py

Update order.py and restaurant.py to import the collections.abc.Iterable module. This ensures compatibility with the latest version of Python.

WassCodeur committed:
Merge remote-tracking branch 'upstream/main' into database

WassCodeur committed:
chore: Update order.py and restaurant.py to import collections.abc.Iterable module

WassCodeur committed:
Refactor order.py and restaurant.py to use collections.abc.Iterable module

Vinyzu commented on README.md:
````diff
@@ -29,3 +30,4 @@ the information overload!
    ```shell
    python -m sincere_singularities
     ```
+2.  Test the bot in your discord server.
````
We will rework the ReadMe in the Presentation phase anyways.

Vinyzu reviewed (approved):
LGTM for now.
Will do some refractoring of the code tomorrow but its important that we get this code in the main branch to implement it tomorrow.

koviubi56 reviewed (approved):
LGTM
